ROCANA-10050: Kite exposes which partitions were updated when a writer is closed

joey opened the pull request.

joey committed:
ROCANA-10050: Kite exposes which partitions were updated when a writer is closed

matthew-dailey commented on kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/PartitionedDatasetWriter.java:
```diff
@@ -61,6 +63,9 @@
   private final Map<String, Object> provided;
 
+  private final List<PartitionUpdateListener> listeners;
+  protected final Set<String> updatedPartitions;
```
nit: should the member `updatedPartitions` have a different name from `updatedPartitions` in its inner classes to avoid any ambiguity?  I know they're static inner classes and can't reference this member, but it may help with readability.

joey replied:
🤔

matthew-dailey commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/TestPartitionUpdateReporter.java:
```diff
@@ -0,0 +1,92 @@
+          updatedPartitions.add(partition);
+        }
+      });
+    }
```
nit: should the test fail on `else`?  The test will fail further down when `updatedPartitions` is empty, but it may be hard to figure out the reason is because no `PartitionUpdateListener` was registered

matthew-dailey commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/TestPartitionUpdateReporter.java:
```diff
@@ -0,0 +1,92 @@
+        @Override
+        public void partitionUpdated(String partition) {
+          updatedPartitions.add(partition);
+        }
```
What we want is One write operation corresponds to Zero-to-many partitions updated, right?  Should `partitionUpdated` accept a `Collection<String>`instead?  This test implementation (appending to one collection) doesn't isolate where updated partitions come from: e.g. they could come from multiple calls to `writer.write()`, maybe from different threads.  Does that matter?

Related to that, the code calling `writer.write()` needs to be in charge of clearing out `updatedPartitions` between calls to `writer.write()`, which could easily be forgotten.

joey replied:
@matthew-dailey not quite. What we want to know is which partitions were updated when we close a writer. The way we track this is that partitioned dataset writers actually wrap a set of writers, one per partition. When we open those per-partition writers, we track the partition that was open and then when we close the `DatasetWriter` we send a notification, one per partition. The reason for making a call per partition is if in the future we wanted to, we could do the notification when a per-partition writer was closed instead of the whole `DatasetWriter` so I wanted to leave the interface open to that.

As far as threading goes, for this test since I'm making an anonymous inner class for the listener and a `final` local for the set of updated partitions, I know that any notifications came from this thread because the listener is attached to the new `DatasetWriter` I created, also in this method, so even if you ran JUnit with multiple threads, I won't see any false positives here.

matthew-dailey replied:
Cool, this approach sounds really good then.  So then after [these lines](https://github.com/scalingdata/rocana/blob/rocana-2.1.0/rocana-hdfs-consumer/src/main/java/com/rocana/consumer/hdfs/FileSystemWriter.java#L53-L56) in the HDFS Consumer, we would notify the new DLM Refresh service with the set of updated partitions.  Or maybe it would be at commit time?

joey replied:
It'd be at commit time, so after this line: https://github.com/scalingdata/rocana/blob/rocana-2.1.0/rocana-hdfs-consumer/src/main/java/com/rocana/consumer/hdfs/FileSystemWriter.java#L40

matthew-dailey commented on pom.xml:
```diff
@@ -208,7 +208,7 @@
     <solr.hdp.expected.version>5.2.1</solr.hdp.expected.version> <!-- sanity check to verify we actually run against the expected version rather than some outdated version -->
     <vers.hdp.tika>1.7</vers.hdp.tika>
-    <vers.hdp.jackson>2.3.1</vers.hdp.jackson>
+    <vers.hdp.jackson>2.7.4</vers.hdp.jackson>
```
Are these version changes needed? Rocana uses jackson 1.9.13, so this doesn't seem to be for syncing with that repo

joey replied:
Good question. So, there are two implementations/versions of jackson and for some reason, we use both. This version is the faster XML version so this update is to harmonize with that version in Rocana Ops:
```xml
    <!-- Faster XML version -->
    <version.fasterxml-jackson>2.7.4</version.fasterxml-jackson>
    <!-- ... -->
    <!-- Other jackson -->
    <version.jackson>1.9.13</version.jackson>
```

matthew-dailey reviewed (commented):
This looks really good @joey!  I had some questions about how we would make an implementation of PartitionUpdateListener, along with some other nits.

joey committed:
Fixed missing imports in PartitionedDatasetWriter

* Used class-specific names for the updated partitions sets to make it
  more clear what their scope is
* Added an else clause to catch the case where the partition update
  reporter test doesn't get an instance of a PartitionUpdateReporter.
  This should only happen if Kite internals change, but this will make
  it clear why the test failed.
* Updated the test to write to two partitions instead of just one.

matthew-dailey commented on kite-data/kite-data-core/src/test/java/org/kitesdk/data/TestPartitionUpdateReporter.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    Assert.assertEquals("Unexpected number of updated partitions", 2, updatedPartitions.size());
+    Assert.assertTrue(updatedPartitions.contains("username_hash=1"));
```
nit: an assertion message would be nice here.
nit: could also add an assertion for the second record's partition being present

matthew-dailey reviewed (approved):
+1 @joey!  Feel free to merge without addressing my new nits

joey committed:
Added a check for the other partition in the set
